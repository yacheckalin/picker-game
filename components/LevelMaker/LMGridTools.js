import React, { useRef, useEffect, useState } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import LMGridObjectIcon from "./LMGridObjectIcon";
import { KEYS, DOORS, WALL, WALL_D } from "../../constants";

const StyledContextBlock = styled.div`
  position: absolute;
  top: ${(prop) => prop.top}px;
  left: ${(prop) => prop.left}px;
  animation-name: grid-tool-show-up;
  animation-duration: 1.5s;

  @keyframes grid-tool-show-up {
    from {
      opacity: 0.5;
    }
    to {
      opacity: 1;
    }
  }

  .collection a span {
    width: 20px;
    height: 20px;
  }
`;

const LMGridTools = ({
  open,
  cellWidth,
  cellHeight,
  cellSize,
  parentTop,
  parentLeft,
  gridWidth,
  multipleMode,
  handleInsert,
}) => {
  const gridToolContextMenuRef = useRef();

  const [topOffset, setTopOffset] = useState(parentTop);
  const [leftOffset, setLeftOffset] = useState(parentLeft + cellWidth);
  const [toolWidth, setToolWidth] = useState(0);
  const [toolHeight, setToolHeight] = useState(0);

  useEffect(() => {
    const el = gridToolContextMenuRef.current;
    if (open) {
      setToolWidth(el.offsetWidth);
      setToolHeight(el.offsetHeight);
    }
  });

  const handleAction = (e) => {
    const { key } = e.target.dataset;
    handleInsert({ key });
  };

  return (
    open && (
      <StyledContextBlock
        top={topOffset}
        left={leftOffset}
        ref={gridToolContextMenuRef}
        id="grid-tool-context-container"
      >
        <div className="collection">
          <a
            className={`collection-item btn ${multipleMode && `disabled`}`}
            data-key="GREEN_KEY"
            onClick={handleAction}
          >
            <LMGridObjectIcon tag={KEYS.GREEN_KEY} className={"left"} />
            Green Key
            <span className="secondary-content green">&nbsp;</span>
          </a>
          <a
            className={`collection-item btn ${multipleMode && `disabled`}`}
            data-key="RED_KEY"
            onClick={handleAction}
          >
            <LMGridObjectIcon tag={KEYS.RED_KEY} className={"left"} />
            Red Key
            <span className="secondary-content red">&nbsp;</span>
          </a>
          <a
            className={`collection-item btn ${multipleMode && `disabled`}`}
            data-key="BLUE_KEY"
            onClick={handleAction}
          >
            <LMGridObjectIcon tag={KEYS.BLUE_KEY} className={"left"} />
            Blue Key
            <span className="secondary-content blue">&nbsp;</span>
          </a>
          <a
            className={`collection-item btn ${multipleMode && `disabled`}`}
            data-key="BLUE_DOOR"
            onClick={handleAction}
          >
            <LMGridObjectIcon tag={DOORS.BLUE_DOOR} className={"left"} /> Blue
            Door
            <span className="secondary-content blue darken-3">&nbsp;</span>
          </a>
          <a
            className={`collection-item btn ${multipleMode && `disabled`}`}
            data-key="RED_DOOR"
            onClick={handleAction}
          >
            <LMGridObjectIcon tag={DOORS.RED_DOOR} className={"left"} />
            Red Door
            <span className="secondary-content red darken-3">&nbsp;</span>
          </a>
          <a
            className={`collection-item btn ${multipleMode && `disabled`}`}
            data-key="GREEN_DOOR"
            onClick={handleAction}
          >
            <LMGridObjectIcon tag={DOORS.GREEN_DOOR} className={"left"} />
            Green Door
            <span className="secondary-content green darken-3">&nbsp;</span>
          </a>
          <a
            className="collection-item btn"
            data-key="WALL"
            onClick={handleAction}
          >
            <LMGridObjectIcon tag={WALL} className={`left`} />
            Wall
            <span className="secondary-content indigo">&nbsp;</span>
          </a>
          <a
            className="collection-item btn"
            data-key="WALL_D"
            onClick={handleAction}
          >
            <LMGridObjectIcon tag={WALL_D} className={`left`} />
            Wall Near Door
            <span className="secondary-content indigo lighten-3">&nbsp;</span>
          </a>
        </div>
      </StyledContextBlock>
    )
  );
};

LMGridTools.propTypes = {
  open: PropTypes.bool.isRequired,
  cellWidth: PropTypes.number.isRequired,
  cellHeight: PropTypes.number.isRequired,
  cellSize: PropTypes.number.isRequired,
  parentTop: PropTypes.number.isRequired, // child top offset
  parentLeft: PropTypes.number.isRequired, // child left offset
  gridWidth: PropTypes.number.isRequired,
  handleInsert: PropTypes.func.isRequired,
};

export default LMGridTools;
