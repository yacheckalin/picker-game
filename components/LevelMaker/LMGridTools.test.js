import React from "react";
import { shallow, mount } from "enzyme";
import toJSON from "enzyme-to-json";

import LMGridTools from "./LMGridTools";

describe("<LMGridTools />", () => {
  const props = {
    open: true,
    cellWidth: 20,
    cellHeight: 20,
    cellSize: 3,
    parentTop: 0,
    parentLeft: 0,
    gridWidth: 300,
    multipleMode: false,
    handleInsert: jest.fn(),
  };
  test("renders and displays properly", () => {
    const wrapper = shallow(<LMGridTools {...props} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  test("all tools renders properly", () => {
    const wrapper = shallow(<LMGridTools {...props} />);
    expect(wrapper.find("a")).toHaveLength(8);
  });

  test("multiple select possible only for walls", () => {
    const wrapper = shallow(<LMGridTools {...props} multipleMode={true} />);
    expect(wrapper.find("a.disabled")).toHaveLength(6);

    wrapper.unmount();
  });

  //TODO: Add e2e tests for elements on a grid (by color classes)
});
