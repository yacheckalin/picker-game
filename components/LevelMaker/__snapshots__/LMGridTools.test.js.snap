// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LMGridTools /> renders and displays properly 1`] = `
<styled.div
  id="grid-tool-context-container"
  left={20}
  top={0}
>
  <div
    className="collection"
  >
    <a
      className="collection-item btn false"
      data-key="GREEN_KEY"
      onClick={[Function]}
    >
      <LMGridObjectIcon
        className="left"
        tag={
          Array [
            21,
            "GREEN",
          ]
        }
      />
      Green Key
      <span
        className="secondary-content green"
      >
         
      </span>
    </a>
    <a
      className="collection-item btn false"
      data-key="RED_KEY"
      onClick={[Function]}
    >
      <LMGridObjectIcon
        className="left"
        tag={
          Array [
            22,
            "RED",
          ]
        }
      />
      Red Key
      <span
        className="secondary-content red"
      >
         
      </span>
    </a>
    <a
      className="collection-item btn false"
      data-key="BLUE_KEY"
      onClick={[Function]}
    >
      <LMGridObjectIcon
        className="left"
        tag={
          Array [
            20,
            "BLUE",
          ]
        }
      />
      Blue Key
      <span
        className="secondary-content blue"
      >
         
      </span>
    </a>
    <a
      className="collection-item btn false"
      data-key="BLUE_DOOR"
      onClick={[Function]}
    >
      <LMGridObjectIcon
        className="left"
        tag={
          Array [
            90,
            "BLUE",
          ]
        }
      />
       Blue Door
      <span
        className="secondary-content blue darken-3"
      >
         
      </span>
    </a>
    <a
      className="collection-item btn false"
      data-key="RED_DOOR"
      onClick={[Function]}
    >
      <LMGridObjectIcon
        className="left"
        tag={
          Array [
            92,
            "RED",
          ]
        }
      />
      Red Door
      <span
        className="secondary-content red darken-3"
      >
         
      </span>
    </a>
    <a
      className="collection-item btn false"
      data-key="GREEN_DOOR"
      onClick={[Function]}
    >
      <LMGridObjectIcon
        className="left"
        tag={
          Array [
            91,
            "GREEN",
          ]
        }
      />
      Green Door
      <span
        className="secondary-content green darken-3"
      >
         
      </span>
    </a>
    <a
      className="collection-item btn"
      data-key="WALL"
      onClick={[Function]}
    >
      <LMGridObjectIcon
        className="left"
        tag={1}
      />
      Wall
      <span
        className="secondary-content indigo"
      >
         
      </span>
    </a>
    <a
      className="collection-item btn"
      data-key="WALL_D"
      onClick={[Function]}
    >
      <LMGridObjectIcon
        className="left"
        tag={2}
      />
      Wall Near Door
      <span
        className="secondary-content indigo lighten-3"
      >
         
      </span>
    </a>
  </div>
</styled.div>
`;
